src/utils.rs

```diff
@@ -15,6 +19,92 @@ pub fn clone_request(request: &mut Request) -> Result<Request> {
     Ok(request.try_clone().unwrap()) // guaranteed to not fail if body is already buffered
 }
 
+#[derive(Debug)]
+pub enum CompressionType {
+    Gzip,
+    Deflate,
+    Brotli,
+}
+
+impl FromStr for CompressionType {
+    type Err = anyhow::Error;
+    fn from_str(value: &str) -> anyhow::Result<CompressionType> {
+        match value {
+            "gzip" => Ok(CompressionType::Gzip),
+            "deflate" => Ok(CompressionType::Deflate),
+            "br" => Ok(CompressionType::Brotli),
+            _ => Err(anyhow::anyhow!("unknown compression type")),
+        }
+    }
+}
+
+// See https://github.com/seanmonstar/reqwest/blob/9bd4e90ec3401c2c5bc435c58954f3d52ab53e99/src/async_impl/decoder.rs#L150
+pub fn get_compression_type(headers: &HeaderMap) -> Option<CompressionType> {
+    let mut compression_type = headers
+        .get_all(CONTENT_ENCODING)
+        .iter()
+        .find_map(|value| value.to_str().ok().and_then(|value| value.parse().ok()));
+
+    if compression_type.is_none() {
+        compression_type = headers
+            .get_all(TRANSFER_ENCODING)
+            .iter()
+            .find_map(|value| value.to_str().ok().and_then(|value| value.parse().ok()));
+    }
+
+    if compression_type.is_some() {
+        if let Some(content_length) = headers.get(CONTENT_LENGTH) {
+            if content_length == "0" {
+                return None;
+            }
+        }
+    }
+
+    compression_type
+}
+
+enum Decoder<R>
+where
+    R: Read,
+{
+    PlainText(R),
+    Gzip(GzDecoder<R>),
+    Deflate(ZlibDecoder<R>),
+    Brotli(BrotliDecoder<R>),
+}
+
+impl<R> Read for Decoder<R>
+where
+    R: Read,
+{
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        match self {
+            Decoder::PlainText(decoder) => decoder.read(buf),
+            Decoder::Gzip(decoder) => decoder.read(buf).map_err(|e| {
```

Comment:
To tell the difference between a decompression error and an I/O error this scheme would be possible:

- Add another wrapper between `GzDecoder` and `R`
- If `R` returns an error, set a `bool` field in that wrapper
- If `GzDecoder` returns an error, use [`get_ref()`](https://docs.rs/flate2/latest/flate2/read/struct.GzDecoder.html#method.get_ref) to check that field
  - If it's `false`, the error happened while decompressing

I don't know if that's worth the effort.